coins/monero/src/wallet/seed/classic/mod.rs

```diff
@@ -0,0 +1,250 @@
+use std::collections::HashMap;
+
+use crc::{Crc, CRC_32_ISO_HDLC};
+use rand_core::OsRng;
+
+use crate::random_scalar;
+
+use super::{SeedError, LanguageName};
+
+mod languages;
+
+pub const CLASSIC_SEED_LENGTH: usize = 24;
+pub const CLASSIC_SEED_LENGTH_WITH_CHECKSUM: usize = 25;
+const LANGUAGE_WORD_COUNT: usize = 1626;
+
+#[derive(Clone)]
+pub struct Language {
+  word_list: Vec<String>,
+  word_map: HashMap<String, usize>,
+  trimmed_word_map: HashMap<String, usize>,
+  language_name: LanguageName,
+  unique_prefix_length: usize,
+}
+
+impl Language {
+  /// populates the word_map and trimmed_word_maps
+  fn populate_maps(
+    &mut self,
+    allow_short_words: bool,
+    allow_duplicate_prefix: bool,
+  ) -> Result<(), SeedError> {
+    if self.word_list.len() != LANGUAGE_WORD_COUNT {
+      Err(SeedError::InvalidLanguageWordList)?;
+    }
+
+    for (i, word) in self.word_list.iter().enumerate() {
+      self.word_map.insert(word.clone(), i);
+
+      if word.len() < self.unique_prefix_length && !allow_short_words {
+        Err(SeedError::InvalidWordInLanguage)?;
+      }
+
+      let trimmed = if word.len() > self.unique_prefix_length {
+        utf8_prefix(word, self.unique_prefix_length)
+      } else {
+        word.to_string()
+      };
+
+      if self.trimmed_word_map.contains_key(&trimmed) && !allow_duplicate_prefix {
+        Err(SeedError::DuplicatePrefix)?;
+      }
+
+      self.trimmed_word_map.insert(trimmed, i);
+    }
+
+    Ok(())
+  }
+}
+
+/// returns a new seed for a given lang.
+pub fn new(lang: LanguageName) -> Result<String, SeedError> {
+  let spend = random_scalar(&mut OsRng);
+  bytes_to_words(spend.as_bytes(), lang)
+}
+
+pub fn words_to_bytes(seed: &str) -> Result<[u8; 32], SeedError> {
+  // get seed words
+  let seed_words: Vec<String> = seed.split_whitespace().map(|w| w.to_string()).collect();
+  let seed_length = seed_words.len();
+  if seed_length != CLASSIC_SEED_LENGTH && seed_length != CLASSIC_SEED_LENGTH_WITH_CHECKSUM {
+    Err(SeedError::InvalidSeedLength)?;
+  }
+
+  // find the language
+  let has_checksum = seed_length == CLASSIC_SEED_LENGTH_WITH_CHECKSUM;
+  let (matched_indices, lang) = find_seed_language(&seed_words, has_checksum)?;
+
+  // convert to  bytes
+  let mut result: [u8; 32] = [b'0'; 32];
+  let word_list_len = lang.word_list.len();
+  for i in 0 .. 8 {
+    let mut indices: [usize; 4] = [0; 4];
+    indices[1] = matched_indices[i * 3];
+    indices[2] = matched_indices[i * 3 + 1];
+    indices[3] = matched_indices[i * 3 + 2];
+
+    indices[0] = indices[1] +
```

Comment:
Please use parentheses for all of these cases. It's a mess to read right now, partially due to the mental requirement for OoO, partially because rustfmt doesn't know how to handle it.